[Txn3] Do not spam LWTs on staging values

gmaretic opened the pull request:
**Goals (and why)**:
==COMMIT_MSG==
Transactions_3 now has proper concurrency controls when reading a staging value.
==COMMIT_MSG==

Previously, on reading a staging value in the transactions table, we would do a CAS to touch the value, followed by a PUT. When multiple transactions were trying to read the same value, this would result in concurrent competing CAS (even though they were semantically the same), which could quickly explode and could continue going on even after a successful PUT.

**Implementation Description (bullets)**:
We do the following:
- we ensure only a single touch is in flight at any moment in time for the transactions service
- we coalesce concurrent TOUCH + PUT requests for the same start timestamp
- we cache the result of recent TOUCH + PUT requests, so only a single instance actually causes KVS calls, in absence of failure

This is done through a combination of synchronisation and a loading cache: when we need to do a TOUCH and a PUT in the protocol, the TOUCH is synchronised, preventing parallel CAS on the KVS level. The loading cache coalesces concurrent calls for the same timestamp and, if successful, also avoids future calls for that timestamp by shortcutting to the result of the cache lookup. 
**Testing (What was existing testing like?  What have you done to improve it?)**:
Existing tests + added tests that exercise the behaviour: coalescing of concurrent requests, serialising CASs, not doing unnecessary touches or puts.

**Concerns (what feedback would you like?)**:
Tests are implementation dependent, but they do test an implementation detail, so some amount of it is hard to avoid. I did add some docs warnings if changes are made that might invalidate them. Also, tests now look ugly because they were written before the inter-rim "fix", but we can live with that probably.

This is probably the best we can do short of implementing a batch CAS endpoint in Cassandra. The risk here is that if we manage to read so many distinct staging start timestamps that we cannot serially do the touches fast enough, we will start falling behind. I do not think this is a realistic issue as once the original commit's PUT goes through, it will unblock all further reads for that start timestamp, even if there are touches waiting on the lock, but definitely worth a short discussion.

**Where should we start reviewing?**:
ResilientCommitTimestampPutUnlessExistsTable

**Priority (whenever / two weeks / yesterday)**:
🔥 but we also need to be careful...

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:
```diff
@@ -28,18 +31,36 @@
     private final TwoPhaseEncodingStrategy encodingStrategy;
     private final Supplier<Boolean> acceptStagingReadsAsCommitted;
 
+    private final LoadingCache<CellInfo, Long> touchCache = CacheBuilder.newBuilder()
```
Can we use Caffeine instead of Guava caches? Baseline checkstyle should flag this https://github.com/palantir/gradle-baseline/blob/ca93cc1fd072ebcf00866364ba85f8f0d1d63d20/.baseline/checkstyle/checkstyle.xml

gmaretic replied:
I knew one was preferred, but wasn't sure which one and this was used elsewhere. Thanks for flagging 👍

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:
```diff
@@ -28,18 +31,36 @@
     private final Supplier<Boolean> acceptStagingReadsAsCommitted;
 
+    private final LoadingCache<CellInfo, Long> touchCache = CacheBuilder.newBuilder()
+            .maximumSize(TOUCH_CACHE_SIZE)
```
What happens if concurrent transaction rate exceeds cache size? Do we fall off a cliff and increase puts on cache miss?

gmaretic replied:
We have to read a staging value in the transactions table, then we try to load. While loading is happening all threads that reach this point will be coalesced. Once the value is loaded it is guaranteed that all reads will forever see the committed value. So the only time a cache hit is actually relevant is if we have read a staging value, but were then so slow we don't get coalesced with someone else. If then our result has already been evicted, it's still fine -- we'll try to do a CAS, fail, but verify it's actually committed now and we'll return the value skipping the put

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:
```diff
@@ -79,8 +100,32 @@ public ListenableFuture<Map<Long, Long>> get(Iterable<Long> cells) {
                 MoreExecutors.directExecutor());
     }
 
+    private synchronized FollowUpAction touchAndReturn(CellInfo cellAndValue) {
```
What all is this `synchronized` concurrency controlling? Do we expect contention under load for cache misses?

gmaretic replied:
This is so we don't have parallel CAS requests for different start timestamps. Since they could still be in the same Cassandra row, we'd contend with ourselves

jeremyk-91 replied:
So this means the CASes will queue up and execute serially. I don't think this is the worst outcome (definitely better than on develop, and you have benchmarked this), but still a bit leery of this

gmaretic replied:
We could put it behind a lock based on the row they are actually trying to CAS to. Will also benchmark that

schlosna commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:
```diff
@@ -79,8 +100,32 @@ public ListenableFuture<Map<Long, Long>> get(Iterable<Long> cells) {
+        Cell cell = cellAndValue.cell();
+        byte[] actual = cellAndValue.value();
+        try {
+            store.checkAndTouch(cell, actual);
```
Something I don't quite understand with the transaction v3 protocol in step 4.i is why the `CAS({V, STAGING}, {V, STAGING})` followed by `PUT({V, COMMITTED})` instead of just a single `CAS({V, STAGING}, {V, COMMITTED})`. If we're going to eat the overhead of CAS with Cassandra Paxos, shouldn't we check for staging and set committed? If the CAS succeeds we're done and if a racing worker has already committed, then there's a no reason to PUT.

https://github.com/palantir/atlasdb/blob/develop/docs/source/miscellaneous/doc/adr/0017-transactions3.md

gmaretic replied:
If we do a CAS from staging to committed we are still vulnerable to the same bug as before. We could be in this state:

1. V, Staging
2. W, Staging
3. Nothing

If we do a CAS (V, Staging) to (V, Committed) and it was only partially applied, and then we also do a CAS (W, Staging) to (W, Committed), then we could end up with

1. V, Committed 
2. W, Committed
3. Nothing

Then a read from nodes (1, 3) will return V, Committed, but a later read from a quorum that contains (W, Committed) will return (W, Committed)

gmaretic commented:
As discussed yesterday, these CASes still can contend with the auto-batched PUE. We could also serialise this by plugging it into the same autobatcher and then running calls serially under the hood. We can potentially add some sort of limit on the CASes themselves to make sure PUEs always make progress.

The real fix would be to extend the Cassandra endpoint to be able to batch all the PUEs and CASes together -- how much work do you think that would be, and should we invest in it @jeremyk-91 ?

jeremyk-91 commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableTest.java:
```diff
@@ -17,58 +17,73 @@
+        validating = !(boolean) parameter;
         pueTable = new ResilientCommitTimestampPutUnlessExistsTable(
-                spiedStore, TwoPhaseEncodingStrategy.INSTANCE, ((Supplier<Boolean>) parameter));
+                spiedStore, TwoPhaseEncodingStrategy.INSTANCE, () -> !validating);
```
uh... that's a lot of negatives?

jeremyk-91 commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableTest.java:
```diff
@@ -159,6 +174,160 @@ public void onceNonNullValueIsReturnedItIsAlwaysReturned() {
+        if (validating) {
+            verify(spiedKvs, times(100)).checkAndSet(any());
+        } else {
+            verify(spiedKvs, never()).checkAndSet(any());
```
ah, nice.

jeremyk-91 commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableTest.java:
```diff
@@ -159,6 +174,160 @@ public void onceNonNullValueIsReturnedItIsAlwaysReturned() {
+        } else {
+            verify(spiedKvs, never()).checkAndSet(any());
+        }
+        verify(spiedStore, times(51)).put(anyMap());
```
Might be missing something: this is 51 and not 50, because of the setupStagingValues() call?

gmaretic replied:
yes

jeremyk-91 commented on atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableTest.java:
```diff
@@ -180,24 +349,68 @@ private static Long firstSuccessfulRead(PutUnlessExistsTable<Long, Long> putUnle
+     * {@link KvsConsensusForgettingStore} and {@link ResilientCommitTimestampPutUnlessExistsTable}. If implementation
+     * details are changed, it may invalidate tests relying on this class.
+     */
+    private static class UnreliableKvsConsensusForgettingStore extends KvsConsensusForgettingStore {
```
This is pretty clever!

jeremyk-91 reviewed (commented):
I'm confident that this PR does what it claims to, and the tests are pretty clever.

Should we do the global batching: yeah we can do that, though it'll probably take maybe three days in coding time all things considered, and unfortunately quite a bit more to roll things out globally. Given the severity of the original incident, I think we should give that a serious attempt though, if it's not too onerous (@leonz / @Sam-Kramer for SA: we're likely to need a new Cassandra version with a new Thrift endpoint).

jeremyk-91 reviewed (commented):
I think we're mostly good here. We need to be very careful with the rollout though (given issues seen before).

gmaretic commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:
```diff
@@ -98,30 +176,50 @@ private Map<Long, Long> processReads(Map<Cell, byte[]> reads, Map<Long, Cell> st
+                                    + "Treating staging values as committed for 60 seconds to ensure liveness.",
+                            SafeArg.of("startTs", startTs),
+                            SafeArg.of("commitTs", commitTs),
+                            SafeArg.of("timeTaken", timeTaken));
```
We now only log once an hour. With the metrics on consensus forgetting store, we should be covered

gmaretic commented:
Addressed comments. In particular, I rate limit the log lines to one an hour, but I did not add metrics given we already have these at the `ConsensusForgettingStore` level. If you don't think that is enough, happy to follow-up

jeremyk-91 reviewed (commented):
discussed offline: This is fine, though let's control the rollout - we should get this on atlasdb proxy first, see it's ok, and then roll out to existing txn3 users

svc-autorelease commented:
Released [0.642.0](https://github.com/palantir/atlasdb/releases/tag/0.642.0)
